### DISCUSSION TITLE
knownSize should return Option[Int]

### DISCUSSION THREAD

**Discussion by phrmoy** [scala/scala3#18324] [Feature Requests]
The following code prints `-1`

```
@main def runPlay = {
  println(List(1,2,3).knownSize)
}
```

I was reasonably surprised by this behavior and though this may not (or maybe) qualify as a type safety issue, I walked away with the sense of lack of type safety.

Rationale/Ask: If the length cannot be produced in constant time, then do not return an incorrect value, instead, return `None` since there is nothing to be returned (the value does not exist in constant time effort). If there is some value to be returned, then return a value of type `Some[Int]`.

I searched for `knownSize` and did not find anything related to it.

Thank you.

Pedro

**Comment by nicolasstucki** [comment]
It makes sense to define `knownSize` as an Int for performance reasons. This avoids the creation of new objects for some core operations and may be helpful for the JIT compiler optimizations.

Just like we have `head` and `headOpt`, we could consider having a `konwnSizeOpt` in the library. Scala 3 uses the Scala 2 library for compatibility. This would need to be added to the `2.13.12+` Scala library in https://github.com/scala/scala.

**Reply by phrmoy** [reply]
Thank you for your reply. I understand the performance concern. Now with that said, here are some thoughts:

1] I do like the `knownSizeOpt` idea as hopefully for a newcomer like me to Scala, I will see both options with the assistance of an IDE and might guess that `knownSize` may return something unexpected since there is a `knownSizeOpt`.

2] The comparison between `head`/`headOpt` and `knownSize`/`knownSizeOpt` is not quite the same because `head`/`headOpt` both have "correct" behavior. That is, `head` throws an exception on an empty sequence.

3] Having `knownSizeOpt` is better than not at all even if `knownSize` may return a -1. I would prefer if `knownSize` also threw an exception like `head` does for consistency and accuracy in behavior (but then this may implicate your performance issue again?). If I had the power, I would then create a `_knownSize` or `unsafeKnownSize` (or anything else to alert the user, specially noobies like me) and allow the inconsistent behavior in it.

Just my thoughts. Thank you.

Pedro

**Comment by som-snytt** [comment]
`knownSize` is not a noob-friendly feature. It's a subtle and suitably obscure mechanism to avoid forcing lazy collections. It's performance-sensitive because used for other ops.

The doc could suggest the quaint `sizeIs` DSL for beginners, or perhaps `sizeCompare`, for times when `size` or `length` is inadvisable.

Compare the huge comment on (deprecated) `hasDefiniteSize`.
